services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DefaultS3TransferManager.java

```diff
@@ -30,54 +32,93 @@
+    DefaultS3TransferManager(S3CrtAsyncClient s3CrtAsyncClient, UploadDirectoryManager uploadDirectoryManager) {
         this.s3CrtAsyncClient = s3CrtAsyncClient;
+        this.transferConfiguration = TransferConfiguration.builder().build();
+        this.uploadDirectoryManager = uploadDirectoryManager;
     }
 
     @Override
     public Upload upload(UploadRequest uploadRequest) {
-        PutObjectRequest putObjectRequest = uploadRequest.putObjectRequest();
-        AsyncRequestBody requestBody = requestBodyFor(uploadRequest);
+        try {
+            assertNotObjectLambdaArn(uploadRequest.putObjectRequest().bucket(), "upload");
+
+            PutObjectRequest putObjectRequest = uploadRequest.putObjectRequest();
+            AsyncRequestBody requestBody = requestBodyFor(uploadRequest);
 
-        CompletableFuture<PutObjectResponse> putObjFuture = s3CrtAsyncClient.putObject(putObjectRequest, requestBody);
+            CompletableFuture<PutObjectResponse> putObjFuture = s3CrtAsyncClient.putObject(putObjectRequest, requestBody);
 
-        CompletableFuture<CompletedUpload> future = putObjFuture.thenApply(r -> DefaultCompletedUpload.builder()
-                                                                                                      .response(r)
-                                                                                                      .build());
-        return new DefaultUpload(CompletableFutureUtils.forwardExceptionTo(future, putObjFuture));
+            CompletableFuture<CompletedUpload> future = putObjFuture.thenApply(r -> DefaultCompletedUpload.builder()
+                                                                                                          .response(r)
+                                                                                                          .build());
+            return new DefaultUpload(CompletableFutureUtils.forwardExceptionTo(future, putObjFuture));
+        } catch (Throwable throwable) {
+            return new DefaultUpload(CompletableFutureUtils.failedFuture(throwable));
+        }
     }
 
     @Override
-    public Download download(DownloadRequest downloadRequest) {
-        CompletableFuture<GetObjectResponse> getObjectFuture =
-            s3CrtAsyncClient.getObject(downloadRequest.getObjectRequest(),
-                                       AsyncResponseTransformer.toFile(downloadRequest.destination()));
-        CompletableFuture<CompletedDownload> future =
-            getObjectFuture.thenApply(r -> DefaultCompletedDownload.builder().response(r).build());
+    public UploadDirectory uploadDirectory(UploadDirectoryRequest uploadDirectoryRequest) {
+        try {
+            Path directory = uploadDirectoryRequest.sourceDirectory();
+
+            if (!Files.exists(directory) || Files.isRegularFile(directory)) {
+                throw new IllegalArgumentException("The source directory provided either does not exist or is not a directory");
+            }
```

Comment:
Can we throw different exception messages for these two cases?

Nit: Validate.isTrue for code golf.